File: Cake Distribution Problem[2-May-2023]/solution.py
#  >> PYTHON CODE <<


def maxSweetness(self, sweetness, n, k):
        def good(x,sweetness,k):
            n=len(sweetness)
            s,cnt=0,0
            
            for i in range(0,n):
                s+=sweetness[i]
                if(s>=x):
                    s=0
                    cnt+=1
            
            return cnt>=k+1
            
        low,high=0,1e9
        ans=-1
        
        while(low<=high):
            mid=(low+high)//2
            if(good(mid,sweetness,k)):
                low=mid+1
                ans=mid
            else:
                high=mid-1

        return int(ans)

File: Cake Distribution Problem[2-May-2023]/solution.cpp
//  >> C++ CODE <<


int maxSweetness(vector<int>& sweetness, int n, int k) {
    int low = 0, high = 1e9, ans = -1;

    auto good = [&](int x) -> bool {
        int s = 0, cnt = 0;

        for (int i = 0; i < n; i++) {
            s += sweetness[i];
            if (s >= x) {
                s = 0;
                cnt++;
            }
        }

        return cnt >= k + 1;
    };

    while (low <= high) {
        int mid = (low + high) / 2;
        if (good(mid)) {
            low = mid + 1;
            ans = mid;
        }
        else {
            high = mid - 1;
        }
    }

    return ans;
}